### PR TITLE
Avoid allocations

### PR DESCRIPTION
**Pull request by MiguelCompany**
This is an improvement over #207 that removes allocations on more cases, including rmw_serialize.

I am rebasing #203 on top of this one, so if this one gets merged the other will be easier to review and merge.

It is passing the tests locally, so I hope it can make it into bouncy

### CODE REVIEW

File: rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp [hunk 1]
```diff
@@ -994,13 +997,28 @@ bool TypeSupport<MembersType>::serialize(
...
   assert(payload);
 
-  auto ser = static_cast<eprosima::fastcdr::Cdr *>(data);
-  if (payload->max_size >= ser->getSerializedDataLength()) {
-    payload->length = static_cast<uint32_t>(ser->getSerializedDataLength());
-    payload->encapsulation = ser->endianness() ==
-      eprosima::fastcdr::Cdr::BIG_ENDIANNESS ? CDR_BE : CDR_LE;
-    memcpy(payload->data, ser->getBufferPointer(), ser->getSerializedDataLength());
-    return true;
+  auto ser_data = static_cast<SerializedData *>(data);
```
Comment (Karsten1987) [inline rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp]: The function could be strongly typed then, right? So this cast could go.

File: rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp [hunk 2]
```diff
@@ -994,13 +997,28 @@ bool TypeSupport<MembersType>::serialize(
...
+  if (ser_data->is_cdr_buffer) {
```
Comment (Karsten1987) [inline rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp]: I am a little bit confused. Why would you call `serialize` on already binary data?

File: rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp [hunk 3]
```diff
@@ -1014,21 +1032,41 @@ bool TypeSupport<MembersType>::deserialize(
   assert(data);
   assert(payload);
 
-  auto buffer = static_cast<eprosima::fastcdr::FastBuffer *>(data);
-  if (!buffer->reserve(payload->length)) {
-    return false;
+  auto ser_data = static_cast<SerializedData *>(data);
```
Comment (Karsten1987) [inline rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp]: i think this function also could be strongly typed.

File: rmw_fastrtps_cpp/src/rmw_publish.cpp [hunk 4]
```diff
@@ -45,17 +46,10 @@ rmw_publish(const rmw_publisher_t * publisher, const void * ros_message)
...
-    info->typesupport_identifier_))
-  {
-    RMW_SET_ERROR_MSG("cannot serialize data");
-    return RMW_RET_ERROR;
-  }
-  if (!info->publisher_->write(&ser)) {
+  rmw_fastrtps_cpp::SerializedData data;
+  data.is_cdr_buffer = false;
+  data.data = const_cast<void *>(ros_message);
+  if (!info->publisher_->write(&data)) {
```
Comment (Karsten1987) [inline rmw_fastrtps_cpp/src/rmw_publish.cpp]: shouldn't this be `data.data` ?

**Comment by dirk-thomas** [comment]
Looking at the first commit this doesn't seem to only be an improvement but it seems to also fix something. Can you please clarify what the problem without that fix is with the current code.

**Comment by dirk-thomas** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/4771/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/1636/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/3957/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/4828/)

File: rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp (hunk 1 above)
Comment (MiguelCompany) [reply rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp]: Not possible, as it is the implementation of Fast-RTPS TopicDataType interface, declared in [this line](https://github.com/eProsima/Fast-RTPS/blob/fb91928914f7f8e47d0ef10bae26b207524b75ad/include/fastrtps/TopicDataType.h#L52)

File: rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp (hunk 3 above)
Comment (MiguelCompany) [reply rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp]: Not possible, as it is the implementation of Fast-RTPS TopicDataType interface, declared in [this line](https://github.com/eProsima/Fast-RTPS/blob/fb91928914f7f8e47d0ef10bae26b207524b75ad/include/fastrtps/TopicDataType.h#L60)

File: rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp (hunk 2 above)
Comment (MiguelCompany) [reply rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp]: This is the case when the call comes from rmw_publish_serialized_message.

Comment (Karsten1987) [reply rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp]: so right now, there is no call to serialize. The raw data will simply be wrapped in `Cdr` and directly send.
I feel like, this decision of whether it's actually necessary to serialize or not should be taken outside of the serialize function.

File: rmw_fastrtps_cpp/src/rmw_publish.cpp (hunk 4 above)
Comment (MiguelCompany) [reply rmw_fastrtps_cpp/src/rmw_publish.cpp]: As stated in [this comment](https://github.com/eProsima/rmw_fastrtps/blob/d4f1dd8453154f02f4fa5ddca798dc9b8bd8bf4a/rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport.hpp#L45), it is not. The write method is provided with a pointer to a SerializedData struct, that will in turn be passed on to the serialize method on TypeSupport_impl.

**Comment by dirk-thomas** [comment]
@MiguelCompany since I noticed you commenting on other questions in this ticket can you please respond to the question in [this comment](https://github.com/ros2/rmw_fastrtps/pull/211#issuecomment-398810406) asap?

**Comment by MiguelCompany** [comment]
> Can you please clarify what the problem without that fix is with the current code.

There are some edge cases where the estimated size will be short by up to 4 bytes. That would provoke a reallocation of the internal buffer when serializing the last field on some messages.

File: rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp (hunk 2 above)
Comment (MiguelCompany) [reply rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp]: When you say 'right now', do you mean on the code currently on master, or on this PR.

Whenever data is sent, the serialize function of TypeSupport class is called. In the code currently on master it just copies an externally serialized buffer into the SerializedPayload (that points) to the internal history record buffer.

The code on this PR does that on some cases (namely rmw_publish_serialized) and performs serialization directly on others (rmw_publish). On the latter case we save the allocation of a temporary buffer.

**Pull request merged**